ext/etc/etc.c

```diff
@@ -200,6 +237,56 @@ setup_passwd(struct passwd *pwd)
 			 0		/*dummy*/
 	);
 }
+
+#if HAVE_GETPWUID_R
+struct getpwuid_r_args {
+    uid_t uid;
+    char *buf;
+    size_t bufsize;
+    struct passwd *result;
+    struct passwd pwstore;
+};
+
+# define GETPWUID_R_ARGS(uid_, buf_, bufsize_) (struct getpwuid_r_args) \
+    {.uid = uid_, .buf = buf_, .bufsize = bufsize_, .result = NULL}
+
+static void *
+nogvl_getpwuid_r(void *args)
+{
+    struct getpwuid_r_args *arg = args;
+    return (void *)(VALUE)getpwuid_r(arg->uid, &arg->pwstore, arg->buf, arg->bufsize, &arg->result);
+}
+#endif
+
+# ifdef HAVE_GETPWNAM_R
+struct getpwnam_r_args {
+    char *login;
+    char *buf;
+    size_t bufsize;
+    struct passwd *result;
+    struct passwd pwstore;
+};
+
+# define GETPWNAM_R_ARGS(login_, buf_, bufsize_) (struct getpwnam_r_args) \
+    {.login = login_, .buf = buf_, .bufsize = bufsize_, .result = NULL}
```

Comment:
IIRC, the designated initializer is a C99 feature.

C.f., https://github.com/ruby/etc/tree/nogvl

Reply:
That nogvl tree looks good to me.  If you want to merge it and close this pull request, you have my support.  Thank you for taking the time to fix these issues.